File: leetcode/minimum_size_subarray_sum.js
/* Given an array of n positive integers and a positive integer s, find the 
 * minimal length of a subarray of which the sum >= s. If there isn't one, 
 * return 0 instead.
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */

/**
 * @param {number} s
 * @param {number[]} nums
 * @return {number}
 */
var minSubArrayLen = function(s, nums) {
    var result = null, begin = 0, sum = 0;
    for (var i = 0; i < nums.length; ++i) {
        for (sum += nums[i]; sum >= s; sum -= nums[begin++]) {
            if (!result || i - begin + 1 < result)
                result = i - begin + 1;
        }
    }
    return result ? result : 0;
};

File: lintcode/minimum_size_subarray_sum.py
class Solution:
    # @param nums: a list of integers
    # @param s: an integer
    # @return: an integer representing the minimum size of subarray
    def minimumSize(self, nums, s):
        result, begin, sum = None, 0, 0
        for i, num in enumerate(nums):
            sum += num
            while sum >= s:
                if not result or i - begin + 1 < result: result = i - begin + 1
                sum -= nums[begin]
                begin += 1
        return result if result else -1
